src/components/TableOfContents/index.tsx

```diff
@@ -82,36 +82,35 @@ export default function TableOfContents({children, title}) {
             document
               .querySelectorAll(`a[href="#${header}"]`)
               .forEach((aTag) => {
-                aTag.classList.remove("active");
+                aTag.classList.remove('active');
               });
           });
           if (activeLink >= 0) {
             const activeElement = headers[activeLink];
             document
               .querySelectorAll(`a[href="#${activeElement}"]`)
               .forEach((aTag) => {
-                aTag.classList.add("active");
+                aTag.classList.add('active');
               });
             if (activeElement) {
-              router.replace(
-                window.location.href.split("#")[0],
-                window.location.href.split("#")[0] + "#" + activeElement,
-                {shallow: true, scroll: false},
-              );
+              // Update the url hash without updating the history.
+              history.replaceState(history.state, null, '#' + activeElement);
```

Comment:
Even though Next.js v12 added [changes for checking whether or not `scrollToHash()` should be called](https://github.com/vercel/next.js/pull/31921/files#diff-ad584158cf978a8bf737eec493b9e3b16100bec3e42e89954fd628a8e03f6295), I found that the scrolling was still broken.

This might've been because the there was still a `scrollToHash()` call inside the `componentDidUpdate()`: https://github.com/vercel/next.js/blob/b1a7b888906307d93b1a3b342a96d7b609606ec8/packages/next/client/index.tsx#L149-L151

Running the patch and viewing the output of `yarn build` still had janky scrolling where the page would jump around though. This made me think that there might've been some other changes in Next.js that needed to be fixed.

Instead of using next `router.replace()` here, I am using `history.replaceState()` because it's able to update the url with the updated hash without creating new state: https://developer.mozilla.org/en-US/docs/Web/API/History/replaceState